libcudacxx/include/cuda/std/detail/libcxx/include/__bit_reference

```diff
@@ -22,1272 +20,1354 @@
+  friend class __bit_const_reference<_Cp>;
+  friend class __bit_iterator<_Cp, false>;
 
-    _LIBCUDACXX_INLINE_VISIBILITY
-    __bit_reference& operator=(bool __x) noexcept
+public:
+  using __container = typename _Cp::__self;
+
+  _LIBCUDACXX_HIDE_FROM_ABI _CCCL_CONSTEXPR_CXX14 __bit_reference(const __bit_reference&) = default;
+
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 operator bool() const noexcept
+  {
+    return static_cast<bool>(*__seg_ & __mask_);
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 bool operator~() const noexcept
+  {
+    return !static_cast<bool>(*this);
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 __bit_reference&
+  operator=(bool __x) noexcept
+  {
+    if (__x)
     {
-        if (__x)
-            *__seg_ |= __mask_;
-        else
-            *__seg_ &= ~__mask_;
-        return *this;
+      *__seg_ |= __mask_;
+    }
+    else
+    {
+      *__seg_ &= ~__mask_;
     }
+    return *this;
+  }
+
+#if _LIBCUDACXX_STD_VER >= 2023
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY constexpr const __bit_reference&
+  operator=(bool __x) const noexcept
+  {
+    if (__x)
+    {
+      *__seg_ |= __mask_;
+    }
+    else
+    {
+      *__seg_ &= ~__mask_;
+    }
+    return *this;
+  }
+#endif
 
-    _LIBCUDACXX_INLINE_VISIBILITY
-    __bit_reference& operator=(const __bit_reference& __x) noexcept
-        {return operator=(static_cast<bool>(__x));}
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 __bit_reference&
+  operator=(const __bit_reference& __x) noexcept
+  {
+    return operator=(static_cast<bool>(__x));
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 void flip() noexcept
+  {
+    *__seg_ ^= __mask_;
+  }
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 __bit_iterator<_Cp, false>
+  operator&() const noexcept
+  {
+    return __bit_iterator<_Cp, false>(__seg_, static_cast<unsigned>(_CUDA_VSTD::__libcpp_ctz(__mask_)));
+  }
 
-    _LIBCUDACXX_INLINE_VISIBILITY void flip() noexcept {*__seg_ ^= __mask_;}
-    _LIBCUDACXX_INLINE_VISIBILITY __bit_iterator<_Cp, false> operator&() const noexcept
-        {return __bit_iterator<_Cp, false>(__seg_, static_cast<unsigned>(__libcpp_ctz(__mask_)));}
 private:
-    _LIBCUDACXX_INLINE_VISIBILITY
-    __bit_reference(__storage_pointer __s, __storage_type __m) noexcept
-        : __seg_(__s), __mask_(__m) {}
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 explicit __bit_reference(
+    __storage_pointer __s, __storage_type __m) noexcept
+      : __seg_(__s)
+      , __mask_(__m)
+  {}
 };
 
 template <class _Cp>
 class __bit_reference<_Cp, false>
-{
-};
+{};
 
 template <class _Cp>
-inline _LIBCUDACXX_INLINE_VISIBILITY
-void
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 void
```

Comment:
Question: Do we want to take the opportunity and moderinze the code base?

I was previously of the opinion that we should stay close to libcxx but I am less and less convinced

Reply:
What exactly do you mean here?